From: Y Giridhar Appaji Nag <giridhar@appaji.net>
Date: Tue, 15 Jan 2008 13:01:49 +0530
Subject: Support for variadic macros __VA_ARGS__

Description: Support for variadic macros __VA_ARGS__,
 based on a patch from "Bernhard R. Link" <brlink@debian.org> for #359894
Author: Y Giridhar Appaji Nag <giridhar@appaji.net>
---
This patch header follows DEP-3: http://dep.debian.net/deps/dep3/
---
 src/cpplib.c | 51 ++++++++++++++++++++++++++++++++++-----------------
 1 file changed, 34 insertions(+), 17 deletions(-)

diff --git a/src/cpplib.c b/src/cpplib.c
index b054f61..2e97592 100644
--- a/src/cpplib.c
+++ b/src/cpplib.c
@@ -2202,6 +2202,13 @@ static char rest_extension[] = "...";
 /*@notfunction@*/
 #define REST_EXTENSION_LENGTH	(sizeof (rest_extension) - 1)
 
+/*@-readonlytrans@*/
+static char rest_name[] = "__VA_ARGS__";
+/*:=readonlytrans@*/
+
+/*@notfunction@*/
+#define REST_NAME_LENGTH	(sizeof (rest_name) - 1)
+
 /* Create a DEFINITION node from a #define directive.  Arguments are
    as for do_define.  */
 
@@ -2272,28 +2279,38 @@ create_definition (/*@exposed@*/ char *buf, char *limit,
 				  cstring_fromChars (rest_extension)));
 	  }
 
-	if (!is_idstart[(int) *bp])
+	if (limit - bp > size_toInt (REST_EXTENSION_LENGTH)
+	    && strncmp (rest_extension, bp, REST_EXTENSION_LENGTH) == 0)
 	  {
-	    cppReader_pedwarnLit (pfile,
-			    cstring_makeLiteralTemp ("Invalid character in macro parameter name"));
+	    rest_args = 1;
+	    temp->rest_args = 1;
+	    temp->name = rest_name;
+	    temp->length = REST_NAME_LENGTH;
 	  }
-
-	/* Find the end of the arg name.  */
-	while (is_idchar[(int) *bp])
+	else
 	  {
-	    bp++;
-	    /* do we have a "special" rest-args extension here? */
-	    if (limit - bp > size_toInt (REST_EXTENSION_LENGTH)
-		&& strncmp (rest_extension, bp, REST_EXTENSION_LENGTH) == 0)
-	      {
-		rest_args = 1;
-		temp->rest_args = 1;
-		/*@innerbreak@*/ break;
-	      }
+		if (!is_idstart[(int) *bp])
+		  {
+		    cppReader_pedwarnLit (pfile,
+				    cstring_makeLiteralTemp ("Invalid character in macro parameter name"));
+		  }
+	
+		/* Find the end of the arg name.  */
+		while (is_idchar[(int) *bp])
+		  {
+		    bp++;
+		    /* do we have a "special" rest-args extension here? */
+		    if (limit - bp > size_toInt (REST_EXTENSION_LENGTH)
+			&& strncmp (rest_extension, bp, REST_EXTENSION_LENGTH) == 0)
+		      {
+			rest_args = 1;
+			temp->rest_args = 1;
+			/*@innerbreak@*/ break;
+		      }
+		  }
+		temp->length = size_fromInt (bp - temp->name);
 	  }
 
-	temp->length = size_fromInt (bp - temp->name);
-
 	if (rest_args != 0)
 	  {
 	    bp += REST_EXTENSION_LENGTH;
